File: lab3-cryptography/zad2/src/main/java/org/example/Tools.java
package org.example;

import java.security.SecureRandom;
import java.util.Optional;

public class Tools {
    public static boolean isPrime(long num) {
        if (num < 2) {
            return false;
        }
        if (num == 2) {
            return true;
        }
        if (num % 2 == 0) {
            return false;
        }

        var bound = (long) Math.sqrt(num);

        for (var i = 3; i < bound; i += 2) {
            if (num % i == 0) {
                return false;
            }
        }
        return true;
    }

    public static Optional<Long> gcd(long a, long b) {
        if (a == 0 || b == 0) {
            return Optional.empty();
        }

        a = Math.abs(a);
        b = Math.abs(b);
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return Optional.of(a);
    }

    static long extEuclid(long x, long y) {
        long oldR = x;
        long r = y;
        long oldS = 1;
        long s = 0;
        long oldT = 0;
        long t = 1;
        long temp;

        while (r != 0) {
            long quotient = oldR / r;

            temp = r;
            r = oldR - quotient * temp;
            oldR = temp;

            temp = s;
            s = oldS - quotient * temp;
            oldS = temp;

            temp = t;
            t = oldT - quotient * temp;
            oldT = temp;
        }
        return oldS * oldR;
    }

    static long genKey() {
        SecureRandom random = new SecureRandom();
        long result = 0;
        for (int i = 0; i < 32; i++) {
            result <<= 1;
            result |= random.nextInt(2);
        }

        if (result == 0) {
            return genKey();
        } else {
            return result;
        }
    }
}

File: lab3-cryptography/zad3/src/tools.rs
use num_traits::{PrimInt, Zero};
use rand::{random, Rng, SeedableRng};
use rand::distributions::Uniform;
use rand::rngs::StdRng;

pub fn gcd<T: PrimInt>(mut a: T, mut b: T) -> Option<T> {
    if a.is_zero() || b.is_zero() {
        return None;
    }

    while !b.is_zero() {
        let temp = b;
        b = a % b;
        a = temp;
    }

    Some(a)
}


pub fn ext_euclid<T : PrimInt>(x: T, y: T) -> T {
    let mut old_r = x;
    let mut r = y;
    let mut old_s: T = T::one();
    let mut s: T = T::zero();
    let mut old_t : T = T::zero();
    let mut t: T = T::one();
    let mut temp: T;

    while !r.is_zero() {
        let quotient = old_r / r;

        temp = r;
        r = old_r - quotient * temp;
        old_r = temp;

        temp = s;
        s = old_s - quotient * temp;
        old_s = temp;

        temp = t;
        t = old_t - quotient * temp;
        old_t = temp;
    }

    old_s * old_r
}

fn floor_sqrt<T: PrimInt>(target: T) -> T {
    let mut low: T = T::one();
    let mut high: T = T::one();
    while high * high < target {
        low = high;
        high = high << 1;
    }

    let mut delta: T = low / T::from(2).unwrap();
    while !delta.is_zero() {
        let test: T = low + delta;
        if test * test > target {
            delta = delta >> 1;
        } else {
            low = low + delta;
        }
    }
    low
}

pub fn is_prime<T: PrimInt>(num: T) -> bool {
    if num < T::from(2).unwrap() { return false; }
    if num == T::from(2).unwrap() { return true; }
    if num % T::from(2).unwrap() == T::zero() { return false; }

    let bound: T = floor_sqrt(num);
    let mut i: T = T::from(3).unwrap();
    while i <= bound {
        if num % i == T::zero() { return false };
        i = i + T::from(2).unwrap();
    }
    true
}

pub fn gen_key<T: PrimInt>(bits: usize) -> T {
    let rng = &mut StdRng::from_entropy();
    let dist = Uniform::new_inclusive(0, 1);
    let mut result: T = T::zero();

    for _ in 0..bits - 1 {
        result = result << 1;
        let bit: T = T::from(rng.sample(dist)).unwrap();
        result = result | bit;
    }

    return if result.is_zero() {
        gen_key(bits)
    } else {
        result
    }
}
